ast_tools/passes/cse.py

```diff
@@ -0,0 +1,197 @@
+import ast
+import astor
+from copy import deepcopy
+from collections import Counter
+import typing as tp
+import warnings
+
+from . import Pass
+from . import PASS_ARGS_T
+
+from ast_tools.common import gen_free_prefix, gen_free_name, is_free_name
+from ast_tools.immutable_ast import immutable, mutable
+from ast_tools.stack import SymbolTable
+from ast_tools.transformers.node_replacer import NodeReplacer
+from ast_tools.visitors.node_finder import NodeFinder
+
+__all__ = ['cse']
+
+def _is_leaf_expr(node: ast.expr):
+    assert isinstance(nod, ast.expr)
+
+    return isinstance(node,(
+        ast.Attribute,
+        ast.Constant,
+        ast.Name,
+        ast.NameConstant,
+        ast.Num,
+        ast.Subscript,
+        ))
+
+class ExprFinder(NodeFinder):
+    def _get_key(self, node):
+        if isinstance(node, ast.expr):
+            # Need the immutable value so its comparable
+            return immutable(node)
+        else:
+            return None
+
+class ExprReplacer(NodeReplacer):
+    def _get_key(self, node):
+        if isinstance(node, ast.expr):
+            # Need the immutable value so its comparable
+            return immutable(node)
+        else:
+            return None
```

Comment:
Hmm, I wonder if there's away that these two classes can share this function definition so the code is not repeated? Could use multiple inheritance, or just define the function externally and assign the name inside the class body.  Not sure what the most elegant solution would be.

Reply:
Just noting that this function does not use `self` and could (really, should) be a free helper function.

Reply:
It is used in the inheritance pattern for NodeReplacer, so maybe it can be made a staticmethod.